plugins/thumbnails/server/__init__.py

```diff
@@ -53,13 +54,50 @@ def removeThumbnailLink(event):
             model.save(resource, validate=False)
 
 
+def _onUpload(event):
+    """
+    Thumbnail creation can be requested on file upload by passing a reference field
+    that is a JSON object of the following form:
+
+        {
+          "thumbnail": {
+            "width": 123,
+            "height": 123,
+            "crop": True
+          }
+        }
+
+    At least one of ``width`` or ``height`` must be passed. The ``crop`` parameter is optional.
+    """
+    try:
+        ref = json.loads(event.info.get('reference'))
+    except (ValueError, TypeError):
+        return
+
+    if isinstance(ref, dict) and isinstance(ref.get('thumbnail'), dict):
+        width = max(0, ref['thumbnail'].get('width', 0))
+        height = max(0, ref['thumbnail'].get('height', 0))
+
+        if not width and not height:
+            return
+        if not isinstance(width, int) or not isinstance(height, int):
+            return
+
+        file = event.info['file']
```

Comment:
This is entirely theoretical, but I'd add the check of whether `'itemId'` is in the `file` to the top of this method. If it fails, we can skip the time spent parsing attached JSON, which could take relatively longer.